From 06a1fdf214cb72f09a3423e996872cc5b7412bdf Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:20:30 -0400
Subject: [fs] nfsd: add a list_head arg to nfsd_foreach_client_lock

Message-id: <1410898849-22126-147-git-send-email-bfields@redhat.com>
Patchwork-id: 95801
O-Subject: [Patch RHEL7 146/165] nfsd: add a list_head arg to nfsd_foreach_client_lock
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Jeff Layton <jlayton@primarydata.com>

In a later patch, we'll want to collect the locks onto a list for later
destruction. If "func" is defined and "collect" is defined, then we'll
add the lock stateid to the list.

Signed-off-by: Jeff Layton <jlayton@primarydata.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: 3738d50e7f6d04dd58d219cf9111bf927c17c6f2
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4state.c b/fs/nfsd/nfs4state.c
index 226d89e..b661294 100644
--- a/fs/nfsd/nfs4state.c
+++ b/fs/nfsd/nfs4state.c
@@ -5811,6 +5811,7 @@ static void nfsd_print_count(struct nfs4_client *clp, unsigned int count,
 }
 
 static u64 nfsd_foreach_client_lock(struct nfs4_client *clp, u64 max,
+        struct list_head *collect,
         void (*func)(struct nfs4_ol_stateid *))
 {
  struct nfs4_openowner *oop;
@@ -5823,8 +5824,12 @@ static u64 nfsd_foreach_client_lock(struct nfs4_client *clp, u64 max,
     &oop->oo_owner.so_stateids, st_perstateowner) {
    list_for_each_entry_safe(lst, lst_next,
      &stp->st_locks, st_locks) {
-    if (func)
+    if (func) {
      func(lst);
+     if (collect)
+      list_add(&lst->st_locks,
+        collect);
+    }
     if (++count == max)
      return count;
    }
@@ -5836,12 +5841,12 @@ static u64 nfsd_foreach_client_lock(struct nfs4_client *clp, u64 max,
 
 u64 nfsd_forget_client_locks(struct nfs4_client *clp, u64 max)
 {
- return nfsd_foreach_client_lock(clp, max, release_lock_stateid);
+ return nfsd_foreach_client_lock(clp, max, NULL, release_lock_stateid);
 }
 
 u64 nfsd_print_client_locks(struct nfs4_client *clp, u64 max)
 {
- u64 count = nfsd_foreach_client_lock(clp, max, NULL);
+ u64 count = nfsd_foreach_client_lock(clp, max, NULL, NULL);
  nfsd_print_count(clp, count, "locked files");
  return count;
 }
-- 
1.7.1